spec/timezone.html

```diff
@@ -5,6 +5,97 @@
   <h1>Temporal.TimeZone Objects</h1>
   <p>A Temporal.TimeZone object is an immutable Object referencing a time zone.</p>
 
+  <emu-clause id="sec-time-zone-names">
+    <h1>Time Zone Names</h1>
+
+    <p>
+      An ECMAScript implementation must support a number of built-in time zones.
+      At a minimum, implementations must support a built-in time zone named *"UTC"*.
+      If the return value of DefaultTimeZone is different from *"UTC"*, then implementations must support that time zone as a built-in time zone as well.
+      In addition, implementations may support any number of other built-in time zones.
+    </p>
+    <p>
+      <emu-xref href="#sec-temporal-timezone-constructor">The `Temporal.TimeZone` constructor</emu-xref>, when called with the name of a built-in time zone as the argument, will return a valid `Temporal.TimeZone` object.
+      When called with any other string, it will throw a *RangeError* exception.
+    </p>
+    <p>
+      Implementations are encouraged to support the Zone and Link names of the IANA Time Zone Database, and to use the best available current and historical information about their offsets from UTC and their daylight saving time rules in calculations.
+    </p>
+    <p>
+      An ECMAScript implementation that includes the ECMA-402 Internationalization API has additional requirements for the built-in time zones that must be supported, as specified in the ECMA-402 specification.
+    </p>
+
+    <emu-note type="editor">
+      <p>These host hooks need to be added to ES2021 <emu-xref href="#sec-host-hooks-summary"></emu-xref> when merged in.</p>
+    </emu-note>
+
+    <emu-clause id="sec-hostisvalidtimezonename" aoid="HostIsValidTimeZoneName">
+      <h1>HostIsValidTimeZoneName ( _timeZone_ )</h1>
+
+      <p>
+        HostIsValidTimeZoneName is a host-defined abstract operation that takes argument _timeZone_ (a String value).
+        It verifies that _timeZone_ represents a valid built-in time zone.
+      </p>
+      <p>
+        The implementation of HostIsValidTimeZoneName must conform to the following requirements:
+      </p>
+      <ul>
+        <li>It must always complete normally (i.e., not return an abrupt completion.)</li>
+        <li>It must always return either *true* or *false*.</li>
+      </ul>
+      <p>
+        The implementation of HostIsValidTimeZoneName should return *true* if _timeZone_, converted to upper case, is equal to one of the built-in time zone names, converted to upper case.
```

Comment:
Should perhaps https://tc39.es/ecma262/#sec-timezoneestring be modified so that any timezone Date is capable of producing, Temporal can recognize?

Reply:
Do you mean that the output of `Date()` should change from e.g. `'Fri Jan 22 2021 11:07:52 GMT-0800 (Pacific Standard Time)'` to `'Fri Jan 22 2021 11:07:52 -08:00[America/Vancouver]'`? That doesn't seem like it would be web compatible.

That does remind me that HostIsValidTimeZoneName should always recognize all offset-only time zone strings, though. I'll make sure that's added.

Reply:
No, that's not what i mean. I mean that if an implementation lacks 402 and supports "America/Vancouver" somehow via Date, it would be nice if it was compelled to also do so for Temporal.

Reply:
I'm not sure I understand what you mean by supporting a time zone via Date, then. You can get a Date in UTC or in the local time zone, but internally it's the same thing as a Temporal.Instant, it doesn't have a time zone. Time zones aren't parsed by `Date.parse()` as far as I can tell. TimeZoneString is only used for output purposes.

Reply:
it's the "local time zone" part i'm interested in. I would prefer it to not be possible that a compliant engine exists where Temporal has fewer timezones than Date (ie, i want every possible timezone `new Date()` can represent to be representable in Temporal, whether 402 is present or not).

Reply:
Sorry, would you mind spelling out what you mean by `new Date()` "representing" a time zone? Date is just a number of milliseconds. I don't understand how the existing language in this PR...

> If the return value of DefaultTimeZone is different from *"UTC"*, then implementations must support that time zone as a built-in time zone as well.

doesn't cover what you are asking for.

Reply:
`new Date().getTimezoneOffset()` returns a local value, not a UTC value, so Date is not, in fact, just a number of milliseconds - it's also got an inherent timezone, which is present in the offset as well as in the `toString` output (which for me says "PST").

Reply:
I'm not sure that's accurate; `date.getTimezoneOffset()` returns [the UTC offset of the host system's current time zone, at the time specified by `date`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTimezoneOffset). This is consistent with Date being a number of milliseconds since the Unix epoch and not having an inherent time zone. If Date held any other state besides a number of milliseconds, it would be apparent from looking at Date's internal slots in [§21.4.5 Properties of Date Instances](https://tc39.es/ecma262/#sec-properties-of-date-instances).

The `toString` output is controlled by [TimeZoneString](https://tc39.es/ecma262/#sec-timezoneestring) as you linked above, and the output string is unfortunately not mentioned to be anything more than implementation-defined, but at least where I've tried it, the time zone name that's output is always the host's current time zone.

We cannot parse strings like "PST" into time zones because they are not mapped unambiguously to actual time zones. "CST", for example, stands for Central Standard Time, Cuba Standard Time (which is currently at the same offset as EST in the US), and China Standard Time. `Date.parse()` does make an effort to parse those abbreviations, but in the implementation that I tried it in, it only understands the US ones (not, for example CET), and `Date.parse()` is not a good thing to emulate anyway.

I think I understand what you are asking for now, but hopefully this clarifies that the extent of what's possible is to require that implementations must have the output of DefaultTimeZone as a builtin time zone.

Reply:
Sorry if i was unclear - I’m not saying the outputs must be observably linked, I’m saying that I’d like whatever conceptual local timezone Dates represent to always also be supported by Temporal.

Reply:
That would require changing the current implementation defined behavior to be standard, wouldn't it? Could that have web incompatible implications.

Reply:
Thanks. As far as I understand, this PR ensures that.

Reply:
Can you help me understand how? Without spec steps or normative prose dictating that as a requirement, it seems like an engine would be free to only support UTC on Temporal, but continue to have local timezones for Date.

Reply:
Sure! :smile: Maybe I should have said "this PR ensures that to the extent that I believe it to be possible". Since Date does not have a time zone, we should examine the places where you can get information about the host's local time zone through Date. Those are, as you identified, `date.getTimezoneOffset()` (where you can obtain the offset at the time given by the Date) and `date.toString()`/`date.toTimeString()` (where you can additionally obtain an unofficial name of the time zone.)

I believe the constraint is satisfied due to the language in DefaultTimeZone

> The DefaultTimeZone abstract operation returns a String value representing the valid and canonicalized time zone name for the host environment's current time zone.

and the language in the first section

> If the return value of DefaultTimeZone is different from **"UTC"**, and does not satisfy the syntax of a TimeZoneNumericUTCOffset, then implementations must support that time zone as a built-in time zone as well.

So, if the host environment's current time zone is not UTC, and not a constant-offset time zone like `"-08:00"` which some environments might want to provide in order to avoid leaking information about the host, then 262 must support it as a built-in time zone. I believe that this precludes implementations from having a local time zone for Date but not for Temporal.

The unofficial name of the local time zone in [`date.toString()`](https://tc39.es/ecma262/#sec-date.prototype.tostring) ultimately comes from step 8 of [TimeZoneString](https://tc39.es/ecma262/#sec-timezoneestring) which is an implementation-defined string in a form which is not (and cannot be) known to Temporal for reasons that I explained above. Changing this to use an official time zone name would not be web compatible.

The offset from [`date.getTimezoneOffset()`](https://tc39.es/ecma262/#sec-date.prototype.gettimezoneoffset) comes from _t_ − [LocalTime](https://tc39.es/ecma262/#sec-localtime)(_t_) which boils down to −[LocalTZA](https://tc39.es/ecma262/#sec-local-time-zone-adjustment)(_t_, **true**) (and this is also where `toString()` gets its offset.) As per the text of LocalTZA, it is

> an implementation-defined algorithm that returns an integral Number representing the local time zone adjustment, or offset, in milliseconds. The local political rules for standard time and daylight saving time in effect at _t_ should be used to determine the result in the way specified in this section.

We could maybe change this to something like "representing the adjustment, or offset, in milliseconds, of the host environment's current time zone" to make sure the same term is used, but I don't believe the current language leaves open the possibility that a host could choose to have one current time zone in LocalTZA and a different one in DefaultTimeZone.

Another additional restriction that we could add, would be that DefaultTimeZone must return a named or offset time zone for which the offset at all possible times _t_ is equal to the return value of LocalTZA(_t_, **true**). That seems redundant to me, but if you're not convinced by the status quo, then maybe this change would be an improvement?

I hope this helps clear up any doubts. If you have concrete suggestions for tightening the language or if there are loopholes that I haven't noticed, please let me know; otherwise, I will consider this resolved.